lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -130,27 +130,31 @@ class window.Turbolinks
   @loadPage: (url, xhr, options = {}) ->
     triggerEvent 'page:receive'
     options.updatePushState ?= true
-    if upstreamDocument = processResponse(xhr, options.partialReplace)
+    if upstreamDocument = processResponse(xhr)
       reflectNewUrl url if options.updatePushState
 
-      new TurboHead(document, upstreamDocument).update(
-        onHeadUpdateSuccess = ->
-          nodes = changePage(
-            upstreamDocument.querySelector('title')?.textContent,
-            removeNoscriptTags(upstreamDocument.querySelector('body')),
-            CSRFToken.get(upstreamDocument).token,
-            'runScripts',
-            options
-          )
-          reflectRedirectedUrl(xhr) if options.updatePushState
-          options.onLoadFunction?()
-          triggerEvent 'page:load', nodes
-        ,
-        onHeadUpdateError = ->
-          Turbolinks.fullPageNavigate(url.absolute)
-      )
-    else
-      Turbolinks.fullPageNavigate(url.absolute)
+      unless options.partialReplace
+        return new TurboHead(document, upstreamDocument).update(
+          onHeadUpdateSuccess = ->
+            updateBody(upstreamDocument, xhr, options)
+          ,
+          onHeadUpdateError = ->
+            Turbolinks.fullPageNavigate(url.absolute)
+        )
+      return updateBody(upstreamDocument, xhr, options)
+    Turbolinks.fullPageNavigate(url.absolute)
```

Comment:
your logic changes LGTM
My only comment would be that I prefer the style of explicit if/else rather than `return` to terminate control flow